src/gui/qgshighlight.cpp

```diff
@@ -168,6 +169,23 @@ void QgsHighlight::setSymbol( QgsSymbol *symbol, const QgsRenderContext &context
     if ( !symbolLayer )
       continue;
 
+    // For geometry generators we need to generate the geometry using the layer's CRS and transform the generated
+    // geometry to the canvas CRS, see issue https://github.com/qgis/QGIS/issues/48439
+    if ( QgsGeometryGeneratorSymbolLayer *generator = dynamic_cast<QgsGeometryGeneratorSymbolLayer *>( symbolLayer ) )
+    {
+      QgsExpressionContext ctx { context.expressionContext() };
+      QgsFeature feature { mFeature };
+      feature.setGeometry( mOriginalGeometry );
+      ctx.setFeature( feature );
+      QgsGeometry geom { QgsExpression( generator->geometryExpression() ).evaluate( &ctx ).value<QgsGeometry>() };
+      const QgsCoordinateTransform ct { mMapCanvas->mapSettings().layerTransform( mLayer ) };
+      if ( ct.isValid() )
+      {
+        geom.transform( ct );
+      }
+      generator->setGeometryExpression( QStringLiteral( "geom_from_wkt('%1')" ).arg( geom.asWkt() ) );
```

Comment:
This looks very inefficient.
Can we somehow preserve the geometry as such?

Reply:
Tried without success.